_includes/tutorials/multiple-event-types/kafka/markup/dev/make-multi-event-app.adoc

```diff
@@ -0,0 +1,147 @@
+ avroProduceConfigs.put(AbstractKafkaSchemaSerDeConfig.AUTO_REGISTER_SCHEMAS, false); <4>
+ avroProduceConfigs.put(AbstractKafkaSchemaSerDeConfig.USE_LATEST_VERSION, true) <5>
+----
+
+<1> Getting the producer from the supplier
+<2> Streaming over the provided collection of Avro events to send
+<3> Creating the `ProducerRecord` instance, note the use of map-like access to get the required field for the key
+<4> Specifying **_to disable_** automatic schema registation
+<5> Setting to use the latest schema
+
+You have a very similar approach with the Avro producer, but take a took at the type at annotation one - it's an abstract class, `SpecificRecordBase` that every Avro generated class inherits from.  Since the schema for the Avro multi-event topic uses a `union` at the top level, you don't know the concrete type.  Since you want to use the customer id as the key you need to access the field in a map-like fashion by using the field name as it exists in the schema.  This is possible because `SpecificRecordBase` implements the `GenericRecord` interface which provides the `get` method for retrieving a field value by name.
+
+But the biggest difference is the configurations you provide to the producer for the Avro serializer. namely disabling automatic schema registration, otherwise it would override the union schema as the latest one.  Additionally since you've set `use.latest.version` to `true` the serializer looks up the latest version, the union schema, and will use that for serialization. https://www.confluent.io/blog/multiple-event-types-in-the-same-kafka-topic/#avro-unions-with-schema-references[This blog post] by Robert Yokota explains this mechanism in detail.
+
+Next let's move on to the `KafkaConsumer` instances, starting with the Protobuf version.
+
+[source, java]
+.KafkaConsumer for multiple events in Protobuf
+----
+   consumerRecords.forEach(consumerRec -> {
+    CustomerEventProto.CustomerEvent customerEvent = consumerRec.value();
+    switch (customerEvent.getActionCase()) { <1>
+        case PURCHASE:
+            eventTracker.add(customerEvent.getPurchase().getItem()); <2>
+            break;
+        case PAGE_VIEW:
+            eventTracker.add(customerEvent.getPageView().getUrl()); <3>
+            break;
+
+
+
+// details left out for clarity
+----
+
+<1> Using a `switch` statement for the different `enum` types
+<2> Adding the purchased item to the event tracker
+<3> Adding the page view link to the event tracker
+
+With Protobuf when you have a `oneof` field, it generates an `enum` for each message that could possibly be in the field.  Determining which record to work with is straight forward by using different code blocks for each case.  The method to retrieve the `enum` alwasy follows the naming convention get<FieldName>Case.  So in this case the field name is `action` you use `getActionCase` for determining the type in the field.
```

Comment:
I think the last 2 lines are a bit confusing, maybe could be polished a bit for clarity?